tool/cmd/kitex/versions/dependencies.go

```diff
@@ -0,0 +1,266 @@
+)
+
+type CheckResult struct {
+	dep *Dependency
+	err error
+}
+
+// Dependency returns the corresponding Dependency of CheckResult
+func (cr *CheckResult) Dependency() Dependency {
+	return *cr.dep
+}
+
+// Err returns the error in checking process
+func (cr *CheckResult) Err() error {
+	return cr.err
+}
+
+// GoModVersion returns the version of the dependency in the go.mod
+func (cr *CheckResult) GoModVersion() string {
+	if cr.dep == nil {
+		return ""
+	}
+	return cr.dep.goModStr
+}
+
+func (d *Dependency) init() error {
+	if d.RefPath == "" {
+		return errors.New("empty RefPath")
+	}
+
+	if d.MinimalVersion == "" {
+		return errors.New("empty MinimalVersion")
+	}
+	minVer, err := newVersion(d.MinimalVersion)
+	if err != nil {
+		return err
+	}
+	d.minVer = minVer
+
+	return nil
+}
+
+func (d *Dependency) check() error {
+	res, err := runCommand("go list -m " + d.RefPath)
+	if err != nil {
+		return ErrDependencyNotFound
+	}
+
+	verStr := parseGoListVersion(res)
+	// replace with local repository
+	if verStr == "" {
+		return nil
+	}
+
+	d.goModStr = verStr
+	ver, err := newVersion(verStr)
+	if err != nil {
+		return ErrDependencyVersionNotSemantic
+	}
+	if ver.lessThan(d.minVer) {
+		return ErrDependencyVersionNotCompatible
+	}
+	return nil
+}
+
+type dependencyManager struct {
+	dependencies map[string]*Dependency
+}
+
+func (dm *dependencyManager) RegisterDependencies(deps []Dependency) error {
+	for _, dep := range deps {
+		newDep := dep
+		if err := newDep.init(); err != nil {
+			return fmt.Errorf("initialize %s Dependency failed: %s", newDep.RefPath, err)
+		}
+		dm.dependencies[newDep.RefPath] = &newDep
+	}
+	return nil
+}
+
+func (dm *dependencyManager) Check() []*CheckResult {
+	var res []*CheckResult
+	for _, dep := range dm.dependencies {
+		err := dep.check()
+		res = append(res, &CheckResult{dep, err})
+	}
+	return res
+}
+
+func (dm *dependencyManager) clear() {
+	dm.dependencies = make(map[string]*Dependency)
+}
+
+func runCommand(input string) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	arr := strings.Split(input, " ")
+	var cmd *exec.Cmd
+	if len(arr) > 1 {
+		cmd = exec.CommandContext(ctx, arr[0], arr[1:]...)
+	} else {
+		cmd = exec.CommandContext(ctx, arr[0])
+	}
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", err
+	}
+	result := strings.TrimSpace(string(output))
+	return result, nil
+}
+
+var goListVersionRegexp = regexp.MustCompile(`^(\S+)\s+(\S+)(\s+=>\s+)?(\S+)?(\s+)?(\S+)?$`)
+
+// parseGoToolVersion parses the version string returned by `go list -m {dependency}`
+// and returns the exact version of the dependency.
+// e.g.
+//
+//	verStr: "github.com/cloudwego/kitex v0.9.0"
+//	result: "v0.9.0"
+//	verStr: github.com/cloudwego/kitex v0.9.0 => github.com/cloudwego/kitex v0.9.1
+//	result: "v0.9.1"
+//	verStr: "github.com/cloudwego/kitex v0.9.0 => ./kitex"
+//	result: ""
```

Comment:
What's our policy on replacement to local dir?

Reply:
we would ignore this dependency.
Users decide to replace this dependency with the local version which we know nothing about. Checking compatibility is their responsibility.